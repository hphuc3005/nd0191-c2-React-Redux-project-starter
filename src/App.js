import './App.css';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { renderRoutes } from './routes';
import { NavBar } from './components/nav/NavBar';
import { LogInPage } from './pages/LogInPage';
import { NotFoundPage } from './pages/NotFoundPage';

function App() {
  const unauthedRoutes = (
    <>
      <Route path='/login' element={<LogInPage />} />
      <Route path='*' element={<NotFoundPage />} />
    </>
  )

  return (
    <BrowserRouter>
      <NavBar />
      {<Routes>
        {renderRoutes()}
        {unauthedRoutes}
      </Routes>}
    </BrowserRouter>
  );
}

export default App;
