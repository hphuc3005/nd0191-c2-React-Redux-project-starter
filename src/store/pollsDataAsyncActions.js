import { createAsyncThunk } from "@reduxjs/toolkit";
import { checkAuthorization, isLoginSucceeded } from "../helpers/authorization";
import { getAllUsersData, getQuestions, getUserData, saveQuestion, saveQuestionAnswer } from "../helpers/apis";

export const updateUserData = createAsyncThunk(
    "polls-data/updateUserData",
    async (userData) => {
        let response;
        if (userData?.username && userData?.password) {
            response = await isLoginSucceeded(userData);
        } else {
            response = await checkAuthorization({ userid: userData?.username })
        }
        return { userData: response }
    });

export const fetchAllUsers = createAsyncThunk(
    "polls-data/fetchAllUsers",
    async (authorId) => {
        return getAllUsersData().then(res => ({ allUsers: res }))
    });

export const fetchQuestions = createAsyncThunk(
    "polls-data/fetchQuestions",
    async () => {
        return getQuestions().then(
            res => {
                const sortedQuestions = res && Object.values(res).sort((a, b) => {
                    if (a.timestamp === b.timestamp) {
                        return a.author - b.author;
                    }
                    return b.timestamp - a.timestamp;
                })
                return ({ questionsData: sortedQuestions })
            }
        );
    })

export const updateQuestionAnswer = createAsyncThunk(
    "polls-data/updateQuestionAnswer",
    async ({ authedUserId, questionId, answer }) => {
        return saveQuestionAnswer({ authedUserId, questionId, answer }).then(
            async () => ({
                userData: await getUserData(authedUserId).then(res => res),
                questionsData: await getQuestions().then(
                    res => {
                        return res && Object.values(res).sort((a, b) => {
                            if (a.timestamp === b.timestamp) {
                                return a.author - b.author;
                            }
                            return b.timestamp - a.timestamp;
                        })
                    }
                )
            })
        )
    });

export const updateQuestion = createAsyncThunk(
    "polls-data/updateQuestion",
    async ({ optionOneText, optionTwoText, author }) => {
        return saveQuestion({ optionOneText, optionTwoText, author }).then(
            async () => ({
                questionsData: await getQuestions().then(
                    res => {
                        return res && Object.values(res).sort((a, b) => {
                            if (a.timestamp === b.timestamp) {
                                return a.author - b.author;
                            }
                            return b.timestamp - a.timestamp;
                        })
                    }
                )
            })
        )
    });

export const updateLeaderboard = createAsyncThunk(
    "polls-data/updateLeaderboard",
    async () => {
        return getAllUsersData().then((user) => {
            const usersData = Object.values(user).map((user) => {
                return {
                    name: user?.name,
                    id: user?.id,
                    avatarURL: user.avatarURL,
                    answers: (user?.answers && Object.keys(user.answers)?.length) || 0,
                    created: user?.questions?.length || 0,
                };
            });
            usersData.sort((a, b) => {
                if (a.answers === b.answers) {
                    return b.created - a.created;
                }
                return b.answers - a.answers;
            });
            return usersData
        });
    });