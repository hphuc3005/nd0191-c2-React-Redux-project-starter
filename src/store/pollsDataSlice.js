import { createSlice } from "@reduxjs/toolkit";
import {
    fetchAllUsers,
    updateLeaderboard,
    updateQuestion,
    updateQuestionAnswer,
    updateUserData,
} from "./pollsDataAsyncActions";
import { fetchQuestions } from "./pollsDataAsyncActions";

export const initPollsData = {
    userData: {},
    questionsData: [],
    leaderboardData: [],
    allUsers: {},
    isLoading: false,
    error: null,
};

const pollsDataSlice = createSlice({
    name: "polls-data",
    initialState: initPollsData,
    reducers: {
        clearData(state, action) {
            state.userData = { ...initPollsData.userData };
            state.questionsData = { ...initPollsData.questionsData };
            state.leaderboardData = { ...initPollsData.leaderboardData };
            state.allUsers = { ...initPollsData.allUsers };
            state.isLoading = false;
            state.error = null;
        },
    },
    extraReducers(builder) {
        builder
            .addCase(updateUserData.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(updateUserData.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
                state.userData = { error: action.error?.message }
            })
            .addCase(updateUserData.fulfilled, (state, action) => {
                const userData = action.payload.userData || {};
                state.userData = { ...userData };
                state.isLoading = false;
            })
            .addCase(fetchQuestions.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(fetchQuestions.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
            })
            .addCase(fetchQuestions.fulfilled, (state, action) => {
                const questionsData = action.payload.questionsData || {};
                state.questionsData = [...questionsData];
                state.isLoading = false;
            })
            .addCase(updateQuestionAnswer.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(updateQuestionAnswer.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
            })
            .addCase(updateQuestionAnswer.fulfilled, (state, action) => {
                const userData = action.payload.userData || {};
                const questionsData = action.payload.questionsData || {};
                state.userData = { ...userData };
                state.questionsData = [...questionsData];
                state.isLoading = false;
            })
            .addCase(updateQuestion.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(updateQuestion.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
            })
            .addCase(updateQuestion.fulfilled, (state, action) => {
                const questionsData = action.payload.questionsData || {};
                state.questionsData = [...questionsData];
                state.isLoading = false;
            })
            .addCase(updateLeaderboard.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(updateLeaderboard.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
            })
            .addCase(updateLeaderboard.fulfilled, (state, action) => {
                state.leaderboardData = action.payload || [];
                state.isLoading = false;
            })
            .addCase(fetchAllUsers.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(fetchAllUsers.rejected, (state, action) => {
                state.isLoading = false;
                state.error = action.error?.message;
            })
            .addCase(fetchAllUsers.fulfilled, (state, action) => {
                state.allUsers = action.payload.allUsers || {};
                state.isLoading = false;
            });
    },
});

export const pollsDataActions = pollsDataSlice.actions;

export default pollsDataSlice;
