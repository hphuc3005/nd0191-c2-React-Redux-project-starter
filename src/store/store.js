import { configureStore } from "@reduxjs/toolkit";
import logger from "redux-logger";

import pollsDataSlice from "./pollsDataSlice";

export const store = configureStore({
    reducer: {
        pollsData: pollsDataSlice.reducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

