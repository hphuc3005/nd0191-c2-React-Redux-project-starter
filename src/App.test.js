import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';
import '@testing-library/jest-dom'

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { initPollsData } from './store/pollsDataSlice';

test('Render App', () => {
  const initialState = { ...initPollsData };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const view = render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(view).toMatchSnapshot()
  // eslint-disable-next-line testing-library/no-debugging-utils
  // screen.debug();
  const loginForm = screen.getByText(/Login/i);
  const navHome = screen.getByText(/Home/i);
  const navLeaderboard = screen.getByText(/Leaderboard/i);
  const navCreateNewPoll = screen.getByText(/Create New Poll/i);
  expect(loginForm).toBeInTheDocument();
  expect(navHome).toBeInTheDocument();
  expect(navLeaderboard).toBeInTheDocument();
  expect(navCreateNewPoll).toBeInTheDocument();
});
