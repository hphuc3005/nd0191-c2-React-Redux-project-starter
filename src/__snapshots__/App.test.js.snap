// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/add"
            >
              Create New Poll
            </a>
          </li>
        </ul>
      </nav>
      <form
        data-testid="form"
      >
        <h1>
          Login
        </h1>
        <div
          class="login-image"
        >
          <img
            alt="User Avatar"
            src="/images/login_image.jpg"
            style="max-width: 100%; margin-bottom: 10px;"
          />
          <br />
        </div>
        <div>
          <label
            for="username"
            required=""
          >
            Username
          </label>
          <input
            data-testid="username"
            id="username"
            name="username"
            placeholder="Your Username"
            required=""
            type="text"
          />
        </div>
        <div>
          <label
            for="password"
            required=""
          >
            Password
          </label>
          <input
            data-testid="password"
            id="password"
            name="password"
            placeholder="Your Password"
            required=""
            type="password"
          />
        </div>
        <br />
        <button
          data-testid="submit-button"
          id="submit-button"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar"
    >
      <ul
        class="navbar-nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/add"
          >
            Create New Poll
          </a>
        </li>
      </ul>
    </nav>
    <form
      data-testid="form"
    >
      <h1>
        Login
      </h1>
      <div
        class="login-image"
      >
        <img
          alt="User Avatar"
          src="/images/login_image.jpg"
          style="max-width: 100%; margin-bottom: 10px;"
        />
        <br />
      </div>
      <div>
        <label
          for="username"
          required=""
        >
          Username
        </label>
        <input
          data-testid="username"
          id="username"
          name="username"
          placeholder="Your Username"
          required=""
          type="text"
        />
      </div>
      <div>
        <label
          for="password"
          required=""
        >
          Password
        </label>
        <input
          data-testid="password"
          id="password"
          name="password"
          placeholder="Your Password"
          required=""
          type="password"
        />
      </div>
      <br />
      <button
        data-testid="submit-button"
        id="submit-button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
